tests/test_haarpsi.py

```diff
@@ -0,0 +1,60 @@
+    prediction, _ = input_tensors
+    index = haarpsi(prediction.to(device), prediction.to(device), data_range=1., reduction='none')
+    index_255 = haarpsi(prediction.to(device) * 255, prediction.to(device) * 255, data_range=255, reduction='none')
+    assert torch.isclose(index, torch.ones_like(index, device=device)).all(), \
+        f'Expected index to be equal 1, got {index}'
+    assert torch.isclose(index_255, torch.ones_like(index_255, device=device)).all(), \
+        f'Expected index to be equal 1, got {index_255}'
+
+
+def test_haarpsi_symmetry(input_tensors: Tuple[torch.Tensor, torch.Tensor], device: str) -> None:
+    prediction, target = input_tensors
+    result = haarpsi(prediction.to(device), target.to(device), data_range=1., reduction='none')
+    result_sym = haarpsi(target.to(device), prediction.to(device), data_range=1., reduction='none')
+    assert torch.isclose(result_sym, result).all(), f'Expected the same results, got {result} and {result_sym}'
+
+
+def test_haarpsi_zeros_ones_inputs(device: str) -> None:
+    zeros = torch.zeros(1, 3, 256, 256, device=device)
+    ones = torch.ones(1, 3, 256, 256, device=device)
+    haarpsi_ones = haarpsi(ones, ones, data_range=1.)
+    assert torch.isfinite(haarpsi_ones).all(), f'Expected finite value for ones tensos, got {haarpsi_ones}'
+    haarpsi_zeros_ones = haarpsi(zeros, ones, data_range=1.)
+    assert torch.isfinite(haarpsi_zeros_ones).all(), \
+        f'Expected finite value for zeros and ones tensos, got {haarpsi_zeros_ones}'
+
+
+def test_haarpsi_compare_with_matlab(device: str) -> None:
+    prediction = torch.tensor(np.array(Image.open('tests/assets/I01.BMP'))).permute(2, 0, 1)
+    target = torch.tensor(np.array(Image.open('tests/assets/i01_01_5.bmp'))).permute(2, 0, 1)
+    predicted_score = haarpsi(prediction.to(device), target.to(device), data_range=255, reduction='none')
+    target_score = torch.tensor([0.7170]).to(predicted_score)
+    assert torch.isclose(predicted_score, target_score, atol=1e-4),\
+        f'Expected result similar to MATLAB, got diff{predicted_score - target_score}'
+
+
+# ================== Test class: `HaarPSILoss` =================
+def test_haarpsi_loss(input_tensors: Tuple[torch.Tensor, torch.Tensor], device: str) -> None:
+    prediction, target = input_tensors
+    prediction.requires_grad_()
+    loss = HaarPSILoss(data_range=1.)(prediction.to(device), target.to(device))
+    loss.backward()
+    assert prediction.grad is not None, f'Expected finite gradient values after back propagation, got {prediction.grad}'
+
+
+def test_haarpsi_loss_zero_for_equal_input(input_tensors: Tuple[torch.Tensor, torch.Tensor], device: str) -> None:
+    prediction, _ = input_tensors
+    target = prediction.clone()
+    prediction.requires_grad_()
+    loss = HaarPSILoss(data_range=1.)(prediction.to(device), target.to(device))
+    assert torch.isclose(loss, torch.zeros_like(loss), atol=1e-5), \
```

Comment:
Why do we have different precision compared to `test_haarpsi_to_be_one_for_identical_inputs`?

Reply:
Because current implementation of `logit` function is not numerically stable at the borders, and near zero it's has bigger error, than near ones.
Starting from PyTorch 1.7.0 there will be built-in function `torch.logit` which handles values close to 0 and 1 correctly.

Reply:
Current implementation performs the `logit` near `1.` and then is substituted from 1. Therefore there should be the same result, shouldn't it?

Reply:
I noticed, that when passing two tensors with zeros error is bigger, than when passing two tensors with ones. Not sure what exactly causes such behavior.
Original MATLAB and Python versions just fail for such inputs, so having `1e-4` error is not that bad.

Reply:
The difference is due to `rtol=1e05` of the `torch.isclose`.  Resolved.